pkg/service/controller/ingressip/controller_test.go

```diff
@@ -0,0 +1,129 @@
+		var logLevel string
+		flag.StringVar(&logLevel, "logLevel", "5", "test")
+		flag.Lookup("v").Value.Set(logLevel)
+	}
+}
+
+func controllerSetup(startingObjects []runtime.Object, stopChannel chan struct{}, t *testing.T) (*ktestclient.Fake, *watch.FakeWatcher, *IngressIPController) {
+	client := ktestclient.NewSimpleFake(startingObjects...)
+
+	fakeWatch := watch.NewFake()
+	client.PrependWatchReactor("*", ktestclient.DefaultWatchReactor(fakeWatch, nil))
+
+	client.PrependReactor("create", "*", func(action ktestclient.Action) (handled bool, ret runtime.Object, err error) {
+		obj := action.(ktestclient.CreateAction).GetObject()
+		fakeWatch.Add(obj)
+		return true, obj, nil
+	})
+
+	// Ensure that updates the controller makes are passed through to the watcher.
+	client.PrependReactor("update", "*", func(action ktestclient.Action) (handled bool, ret runtime.Object, err error) {
+		obj := action.(ktestclient.CreateAction).GetObject()
+		fakeWatch.Modify(obj)
+		return true, obj, nil
+	})
+
+	_, ipNet, err := net.ParseCIDR("172.16.0.12/28")
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	controller := NewIngressIPController(client, ipNet, 10*time.Minute)
+
+	return client, fakeWatch, controller
+}
+
+func getService(namespace, name string, typeLoadBalancer bool) *kapi.Service {
+	serviceType := kapi.ServiceTypeClusterIP
+	if typeLoadBalancer {
+		serviceType = kapi.ServiceTypeLoadBalancer
+	}
+	return &kapi.Service{
+		ObjectMeta: kapi.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+		},
+		Spec: kapi.ServiceSpec{
+			Type: serviceType,
+		},
+	}
+}
+
+// TestBasicControllerFlow validates controller start, initial sync
+// processing, and post-sync processing.
+func TestBasicControllerFlow(t *testing.T) {
+	configureLogging()
+
+	svc1 := getService("ns", "svc-1", true)
+
+	startingObjects := []runtime.Object{}
+	startingObjects = append(startingObjects, svc1)
+
+	stopChannel := make(chan struct{})
+	defer close(stopChannel)
+
+	_, fakeWatch, controller := controllerSetup(startingObjects, stopChannel, t)
+
+	updated := make(chan bool)
+	deleted := make(chan bool)
+
+	controller.changeHandler = func(change *serviceChange) error {
+		defer func() {
+			if len(change.key) == 0 {
+				deleted <- true
+			} else if change.oldService != nil {
+				updated <- true
+			}
+		}()
+
+		err := controller.processChange(change)
+		if err != nil {
+			t.Errorf("unexpected error: %v", err)
+		}
+
+		return err
+	}
+
+	go controller.Run(stopChannel)
+
+	waitForUpdate := func(msg string) {
+		t.Logf("waiting for: %v", msg)
+		select {
+		case <-updated:
+		case <-time.After(time.Duration(30 * time.Second)):
+			t.Fatalf("failed to see: %v", msg)
+		}
+	}
+
+	waitForUpdate("spec update")
+	waitForUpdate("status update")
+
+	fakeWatch.Delete(svc1)
+
+	t.Log("waiting for the service to be deleted")
+	select {
+	case <-deleted:
+	case <-time.After(time.Duration(30 * time.Second)):
+		t.Fatalf("failed to see expected service deletion")
+	}
```

Comment:
This test is for sanity purposes - the (currently failing) integration test will target the race conditions in question.  Now that the design is mostly shaken out I'm going to ensure better unit test coverage and then work on integration testing.